components/testing/COMPONENT_FPGA_CI_TEST_SHIELD/test_utils.h

```diff
@@ -0,0 +1,470 @@
+            port.status = PortType::StatusPass;
+        }
+        utest_printf("%s\n", port.status == PortType::StatusPass ? "succeeded" : "failed");
+    }
+}
+
+template<typename PortType, typename FunctionType, FunctionType f>
+void test_all_peripherals(std::list<PortType> &matched_ports, std::list<PortType> &not_matched_ports)
+{
+    typedef typename std::list<PortType>::iterator Iter;
+    utest_printf("***Testing all %s peripherals***\n", PortType::PinMap::name);
+
+    if (matched_ports.empty() && not_matched_ports.empty()) {
+        utest_printf("Could not find pins for %s testing \n", PortType::PinMap::name);
+        return;
+    }
+
+    matched_ports.sort(peripheral_less<PortType>);
+    not_matched_ports.sort(peripheral_less<PortType>);
+
+    for (Iter m_it = matched_ports.begin(), nm_it = not_matched_ports.begin();
+            m_it != matched_ports.end() || nm_it != not_matched_ports.end();) {
+        if (m_it != matched_ports.end() && nm_it != not_matched_ports.end()) {
+            if ((*m_it).peripheral < (*nm_it).peripheral) {
+                test_peripheral<PortType, FunctionType, f>(*m_it);
+                ++m_it;
+            } else {
+                test_peripheral<PortType, FunctionType, f>(*nm_it);
+                ++nm_it;
+            }
+        } else if (m_it != matched_ports.end()) {
+            test_peripheral<PortType, FunctionType, f>(*m_it);
+            ++m_it;
+        } else if (nm_it != not_matched_ports.end()) {
+            test_peripheral<PortType, FunctionType, f>(*nm_it);
+            ++nm_it;
+        }
+    }
+}
+
+template<typename PortType, typename FormFactorType, typename PortType::TestFunctionType f>
+void all_ports()
+{
+    std::list<PortType> matched_ports, not_matched_ports;
+    find_ports<PortType, FormFactorType>(matched_ports, not_matched_ports);
+    matched_ports.unique();
+    not_matched_ports.unique();
+    test_all_ports<PortType, FormFactorType, typename PortType::TestFunctionType, f>(matched_ports, not_matched_ports);
+}
+
+template<typename PortType, typename FormFactorType, typename PortType::TestFunctionType f>
+void all_peripherals()
+{
+    std::list<PortType> matched_ports, not_matched_ports;
+    find_ports<PortType, FormFactorType>(matched_ports, not_matched_ports);
+
+    matched_ports.sort(peripheral_less<PortType>);
+    not_matched_ports.sort(peripheral_less<PortType>);
+    matched_ports.unique(peripheral_comparator<PortType>);
+    not_matched_ports.unique(peripheral_comparator<PortType>);
+
+    test_all_peripherals<PortType, typename PortType::TestFunctionType, f>(matched_ports, not_matched_ports);
+}
+
+template<typename PortType, typename FormFactorType, typename PortType::TestFunctionType f>
+void one_peripheral()
+{
+    std::list<PortType> matched_ports, not_matched_ports;
+    find_ports<PortType, FormFactorType>(matched_ports, not_matched_ports);
+
+    utest_printf("***Testing one %s pin configuration***\n", PortType::PinMap::name);
```

Comment:
why do we use `utest_printf` and not stdout?

Reply:
Since this is a test utility shouldn't it use the utest_printf? @maciejbocianski can you comment on this?

Reply:
Exactly, this utility code is intended to be used in tests only,  this is why `utest_printf` was used